### DISCUSSION TITLE
Error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server".

### DISCUSSION THREAD

**Discussion by niallam22** [vercel/next.js#58720] [App Router]
I have created a server action to complete a POST checkout request. I use the server action in my client component, passing the product data to the action.  I have declared the server action with use server but still get this error: Functions cannot be passed directly to Client Components unless you explicitly expose it by marking it with "use server".

I have found this discussion https://github.com/vercel/next.js/discussions/49625
which says it is not possible to use component props directly inside of the action so you have to pass the data into a hidden input and after do the logic inside of the action. However, I am not sure what this means or how to apply it.

Here is my code.

Server action: 
```
'use server'

export async function postCheckout(items: Item[]) {
    const response = await axios.post(`${process.env.NEXT_PUBLIC_API_URL}/checkout`, {
      items: items.map((item) => ({
        productId: item.id,
        quantity: item.quantity,
      })),
    });
}
```

Client component: 
``` 
"use client"

import axios from "axios";
import { postCheckout } from "@/actions/_actions";
import useCart from "@/hooks/use-cart";

const Summary: React.FC<SummaryProps> = () => {
  const items = useCart((state) => state.items);
  const onCheckout = async () => {
      const response = await postCheckout(items)
      window.location = response.data.url;
  };

  return (
      <Button onClick={onCheckout} disabled={items.length === 0} className="w-full mt-6">
        Checkout
      </Button>
  );
};

export default Summary;
```

**Comment by icyJoseph** [comment]
Is `postCheckout` being passed down from a Server Component to the Client Component?

Is the `Client Component` using the "use client" directive?

**Reply by niallam22** [reply]
Yes and yes. The _actions.ts file uses 'use server' and the summary component uses 'use client'. I have updated my initial post which makes that more clear.

**Reply by icyJoseph** [reply]
Ah, this is weird, you are importing `postCheckout`, from the actions into the client component, that should work fine 🤔 I am doing the exact same thing in another project, and it works. Could you share a link to the repository?

**Reply by niallam22** [reply]
Sure here it is - https://github.com/niallam22/ecommerce-store/tree/serverAction

summary component:
https://github.com/niallam22/ecommerce-store/blob/serverAction/app/(routes)/cart/components/summary.tsx

server action: 
https://github.com/niallam22/ecommerce-store/blob/serverAction/actions/_actions.ts


The endpoint it is fetching from this repo:
https://github.com/niallam22/ecommerce-admin/blob/main/app/api/%5BstoreId%5D/checkout/route.ts

**Comment by Nikkakko** [comment]
i have same problem, did you find solution?

**Comment by niallam22** [comment]
no I didn't, please let me know if you do.

**Comment by rferreira98** [comment]
im having this same issue
